lib/app/upload.ex

```diff
@@ -14,53 +14,107 @@ defmodule App.Upload do
+         {:ok, {file_name, upload_response_body}} <-
+           upload_file_to_s3(file_cid, file_extension, image) do
+      # Sample response:
+      # %{
+      #   body: "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n\n
+      #    <CompleteMultipartUploadResult xmlns=\"http://s3.amazonaws.com/doc/2006-03-01/\">
+      #    <Location>https://s3.eu-west-3.amazonaws.com/imgup-original/qvWtbC7WaT.jpg</Location>
+      #    <Bucket>imgup-original</Bucket><Key>qvWtbC7WaT.jpg</Key>
+      #    <ETag>\"4ecd62951576b7e5b4a3e869e5e98a0f-1\"</ETag></CompleteMultipartUploadResult>",
+      #   headers: [
+      #     {"x-amz-id-2",
+      #      "wNTNZKt82vgnOuT1o2Tz8z3gcRzd6wXofYxQmBUkGbBGTpmv1WbwjjGiRAUtOTYIm92bh/VJHhI="},
+      #     {"x-amz-request-id", "QRENBY1MJTQWD7CZ"},
+      #     {"Date", "Tue, 13 Jun 2023 10:22:44 GMT"},
+      #     {"x-amz-expiration",
+      #      "expiry-date=\"Thu, 15 Jun 2023 00:00:00 GMT\", rule-id=\"delete-after-1-day\""},
+      #     {"x-amz-server-side-encryption", "AES256"},
+      #     {"Content-Type", "application/xml"},
+      #     {"Transfer-Encoding", "chunked"},
+      #     {"Server", "AmazonS3"}
+      #   ],
+      #   status_code: 200
+      # }
 
-    # Upload to S3
-    {:ok, body} =
-      image.path
-      |> ExAws.S3.Upload.stream_file()
-      |> ExAws.S3.upload(Application.get_env(:ex_aws, :original_bucket), file_name, acl: :public_read, content_type: image.content_type )
-      |> ExAws.request(get_ex_aws_request_config_override())
+      # Fetch the contents of the returned XML string from `ex_aws`.
+      # This XML is parsed with `sweet_xml`:
+      # github.com/kbrw/sweet_xml#the-x-sigil
+      url = upload_response_body.body |> xpath(~x"//text()") |> List.to_string()
+      compressed_url = "#{@compressed_baseurl}#{file_name}"
+      {:ok, %{url: url, compressed_url: compressed_url}}
+    else
+      {:error, :failure_read} -> {:error, :failure_read}
+      {:error, :invalid_extension_and_cid} -> {:error, :invalid_extension_and_cid}
+      {:error, :invalid_cid} -> {:error, :invalid_cid}
+      {:error, :invalid_extension} -> {:error, :invalid_extension}
+      {:error, :upload_fail} -> {:error, :upload_fail}
+    end
+  end
+
+  defp upload_file_to_s3(file_cid, file_extension, image) do
+    # Creating filename with the retrieved extension
+    file_name = "#{file_cid}.#{file_extension}"
+
+    # Make request.
+    # Return the body of the response if successful.
+    # Otherwise, raise error.
+    try do
+      {:ok, upload_response_body} =
+        image.path
+        |> ExAws.S3.Upload.stream_file()
+        |> ExAws.S3.upload(Application.get_env(:ex_aws, :original_bucket), file_name,
+          acl: :public_read,
+          content_type: image.content_type
+        )
+        |> ExAws.request(get_ex_aws_request_config_override())
+
+      {:ok, {file_name, upload_response_body}}
+    rescue
+      _e ->
+        {:error, :upload_fail}
```

Comment:
This is _exactly_ what we want to avoid doing.
This function doesn't `log` the error so the engineers have no idea what happened.
And then masks the error to the client by defining a `:upload_fail` 
which is not an `{:error, reason}` defined by the `ExAws.S3.upload` library/function.
So when the `client` or consuming function sees this error they have _zero_ insight.